core: stdcm: fix path / travelled path confusion on block availability

eckter opened the pull request:
Well that was a massive bug considering how sneaky it is. It's in `BlockAvailability` (a class that's mocked in most tests), and `Offset<Path> == Offset<TravelledPath>` in most test cases.

`InfraExplorerWithEnvelope.getSimulatedLength()` returned a `Length<Path>`, when it should really be `TravelledPath`. 

I'd add a unit test with stops in `BlockAvailabilityTests`, but that would massively conflict with the other pending PRs. 

Fix some (most?) cases of https://github.com/OpenRailAssociation/osrd/issues/11121.

eckter committed:
core: stdcm: fix path / travelled path confusion on block availability

Signed-off-by: Eloi Charpentier <eloi.charpentier.42@gmail.com>

Erashin commented on core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt:
```diff
@@ -268,12 +272,35 @@ class BlockAvailabilityTests {
+                explorer.getSimulationEndPathOffset(),
+                0.0
+            ) as BlockAvailabilityInterface.Available
+        assertEquals(t - explorer.getFullEnvelope().totalTime, res.maximumDelay)
```
nit: You could also add an assert on the values of getSimulationLength and getSimulationEndPathOffset, just to make sure no one ever switches em again.

eckter replied:
I've added a test on `InfraExplorerWithEnvelope` directly

Erashin reviewed (approved):
Well that must have been fun to debug. Good job, this is huge!
For the test, I'm fine with merging this as is, putting said test in an issue and doing it once the other PRs are merged.

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorerWithEnvelopeImpl.kt:
```diff
@@ -192,12 +193,16 @@ data class InfraExplorerWithEnvelopeImpl(
+    override fun getSimulatedLength(): Length<TravelledPath> {
         if (envelopes.isEmpty()) return Length(0.meters)
         val lastEnvelope = envelopes[envelopes.size - 1]
         return Length(Distance.fromMeters(lastEnvelope.startOffset + lastEnvelope.envelope.endPos))
```
For another PR? 
Can we type the `LocatedEnvelope.startOffset`?

eckter replied:
The envelope framework is still mostly based on untyped `double` and they're not *that* trivial to integer-based `Distances`. The plan was to not migrate it, but we've been dealing with this "legacy" module for a while now...

bougue-pe replied:
OK, I only noted it in the "meta" issue on the matter of offset typing https://github.com/OpenRailAssociation/osrd/issues/10239
Resolving this.

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorerWithEnvelopeImpl.kt:
```diff
@@ -192,12 +193,16 @@ data class InfraExplorerWithEnvelopeImpl(
     }
 
+    override fun getSimulationEndPathOffset(): Offset<Path> {
+        return getIncrementalPath().fromTravelledPath(getSimulatedLength())
```
For another PR? 
Providing an `PathOffsetBuilder` inside `IncrementalPath` would probably be clearer and better mutualization.

eckter replied:
I'm not sure I understand the suggestion, you're saying we should have something like `InfraExplorer.fromTravelPath`? Or build and expose a `PathOffsetBuilder` in `InfraExplorer`?

bougue-pe replied:
I was thinking to attach a `PathOffsetBuilder` to `IncrementalPath` (and remove its `from`/`toTravelledPath` fn), but it may make more sense to attach it to `InfraExplorer`?

eckter replied:
I have to admit I like `IncrementalPath.toTravelledPath` methods. They're the most commonly used methods to convert paths, the `PathOffsetBuilder` is the exception and it's *mostly* local to `ScheduleMetadataExtractor`. We could instantiate a builder inside `IncrementalPath` for code factorization, but incremental paths are instanciated and cloned *a lot*, so the new objects may be expensive.

We could add `InfraExplorer.toTravelledPath` as an utility method (as it's very often chained after `getIncrementalPath` calls), but I'd see it as just wrapping `IncrementalPath.toTravelledPath`.

bougue-pe replied:
OK, perf matters here 🙂 
We could bring `PathOffsetBuilder`'s code a bit closer to what's in `IncrementalPath` to keep the best of both (tiny bit more typing like `IncrementalPath`, tiny bit less indirection like `PathOffsetBuilder`).
It's a detail, and I resolve this.

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt:
```diff
@@ -268,12 +272,35 @@ class BlockAvailabilityTests {
+    @Test
+    fun testNonZeroStartOffset() {
+        val explorer = makeExplorer(5, 5, startOffset = Offset(4_000.meters))
+        val t = 1200.0
```
```suggestion
        val requirementStartTime = 1200.0
```

eckter replied:
Fixed in https://github.com/OpenRailAssociation/osrd/pull/11210/commits/d997b34fb92f0156018bf09a22cd9aa8a87436fa

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt:
```diff
@@ -268,12 +272,35 @@ class BlockAvailabilityTests {
     }
 
+    /**
+     * In this case Offset<Path> != Offset<EnvelopePath>, checks that offsets types aren't mixed up
```
As there is ongoing work around naming and `EnvelopePath` doesn't exist, I think we shouldn't introduce something new.
It looks like current type for envelope sequence is `EnvelopeConcat`.
`EnvelopeSimPath` exists, so it may be used, even though it's not a sequence of designated object.
We may come up with `EnvelopePartPath` to have a sequence of designated object?

Not sure about the meaning here, but maybe just `Offset<Envelope>` ?

eckter replied:
In the current implementation, in every case `Offset<Envelope> == Offset<TravelledPath>`.

But it's true that it's not *necessarily* the case, and we could introduce that type.

There's a slight ambiguity though: for an envelope ranging from position 50 to 100, where does `Offset<Envelope>(50.meters)` point to? Is it the start or the end?

bougue-pe replied:
In my understanding that "`Object`Path" is a sequence of instances of `Object`: `Offset<Envelope>(50.meters)` points to the end, because it's not a path.

Anyway, I think this comment should move a bit, as I still don't understand if it refers to EnvelopeConcat or Envelope:
```suggestion
     * In this case Offset<Path> != Offset<EnvelopeConcat>, checks that offsets types aren't mixed up
```
?

eckter replied:
Oh I missunderstood the initial comment (specifically I didn't read my own code carefully enough).

I intended to write `Offset<TravelledPath>` there, and I agree we can stick to well established types even in comments. Even if envelope offset typing would be nice.

eckter replied:
I've changed it to `TravelledPath`, is it OK for you? I feel like I'm dodging the rest of the discussion, but I don't know if it's still relevant.

bougue-pe replied:
I'm perfectly OK to leave the rest of the discussion to https://github.com/OpenRailAssociation/osrd/issues/10239 (and related work).
Thanks!

bougue-pe reviewed (commented):
Thanks for this bugfix!

Apart from the naming, I guess my first 2 comments can (should?) be tracked a sibling issues of https://github.com/OpenRailAssociation/osrd/issues/10595.
If agreeing on the usefulness of suggestions, I can create issues.

eckter committed:
core: stdcm: avoid exiting too early from main loop

This early return relied on the fact that the nodes
with lowest total time were processed first, which
isn't true anymore

Signed-off-by: Eloi Charpentier <eloi.charpentier.42@gmail.com>

bougue-pe reviewed (approved):
🙏 If you have a unit test for the second commit's bug, it's nice to embed (we can consider that e2e tests are enough, I guess).
